instrumentation/github.com/aws/aws-lambda-go/otellambda/lambda.go

```diff
@@ -0,0 +1,351 @@
+	}
+
+	takesContext, err := validateArguments(handlerType)
+	if err != nil {
+		return errorHandler(err)
+	}
+
+	if err := validateReturns(handlerType); err != nil {
+		return errorHandler(err)
+	}
+
+	// note we will always take context to capture lambda context,
+	// regardless of whether customer takes context
+	if handlerType.NumIn() == 0 || handlerType.NumIn() == 1 && takesContext {
+		return func(ctx context.Context) (interface{}, error) {
+			var temp *interface{}
+			event := reflect.ValueOf(temp)
+			return wrapperInternals(handlerFunc, event, ctx, takesContext)
+		}
+	} else { // customer either takes both context and payload or just payload
+		return func(ctx context.Context, payload interface{}) (interface{}, error) {
+			event, err := payloadToEvent(handlerType.In(handlerType.NumIn()-1), payload)
+			if err != nil {
+				return nil, err
+			}
+			return wrapperInternals(handlerFunc, event.Elem(), ctx, takesContext)
+		}
+	}
+}
+
+// basic implementation of TextMapCarrier
+// which wraps the default map type
+type mapCarrier map[string]string
+
+// Compile time check our mapCarrier implements propagation.TextMapCarrier
+var _ propagation.TextMapCarrier = mapCarrier{}
+
+// Get returns the value associated with the passed key.
+func (mc mapCarrier) Get(key string) string {
+	return mc[key]
+}
+
+// Set stores the key-value pair.
+func (mc mapCarrier) Set(key string, value string) {
+	mc[key] = value
+}
+
+// Keys lists the keys stored in this carrier.
+func (mc mapCarrier) Keys() []string {
+	keys := make([]string, len(mc))
+	i := 0
+	for k := range mc {
+		keys[i] = k
+		i++
+	}
+	return keys
+}
+
+// Adds OTel span surrounding customer handler call
+func wrapper(handlerFunc interface{}) func(ctx context.Context, event interface{}, takesContext bool) []reflect.Value {
+	return func(ctx context.Context, event interface{}, takesContext bool) []reflect.Value {
+
+		ctx, span := tracingBegin(ctx)
+		defer tracingEnd(ctx, span)
+
+		handler := reflect.ValueOf(handlerFunc)
+		var args []reflect.Value
+		if takesContext {
+			args = append(args, reflect.ValueOf(ctx))
+		}
+		if eventExists(event) {
+			args = append(args, reflect.ValueOf(event))
+		}
+
+		response := handler.Call(args)
+
+		return response
+	}
+}
+
+// Determine if an interface{} is nil or the
+// if the reflect.Value of the event is nil
+func eventExists(event interface{}) bool {
+	if event == nil {
+		return false
+	}
+
+	// reflect.Value.isNil() can only be called on
+	// Values of certain Kinds. Unsupported Kinds
+	// will panic rather than return false
+	switch reflect.TypeOf(event).Kind() {
+	case reflect.Chan, reflect.Func, reflect.Map, reflect.Ptr, reflect.UnsafePointer, reflect.Interface, reflect.Slice:
+		return !reflect.ValueOf(event).IsNil()
+	}
+	return true
+}
+
+type wrappedHandler struct {
+	handler lambda.Handler
+}
+
+// Compile time check our Handler implements lambda.Handler
+var _ lambda.Handler = wrappedHandler{}
```

Comment:
There seem to be two different interfaces being implemented here.  They should probably be split into separate files so that it is clear what code goes with which implementation.  The difference between the interfaces should be well documented in the `README`.

Reply:
Moved the two wrappers into separate files. Usage is now shown in doc.go to indicate that the two wrappers are to be used with their respective lambda entry points.